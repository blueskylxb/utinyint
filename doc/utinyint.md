utinyint 0.1.1
=============

Synopsis
--------

    % CREATE EXTENSION utinyint;
    CREATE EXTENSION

    % SELECT 1::utinyint;


Description
-----------

This module provides 1 byte signed integer type with operators,
index access methods and support functions. You can use it as
like the corresponding data type in other RDBMS as well as
smallint, integer or bigint in PostgreSQL. Although the byte
used for 1 utinyint is exactly 1 byte, you need to take account
into its position in the row if you really care the table size
since PostgreSQL usually pads extra bytes for the alignment of 4.

utinyint holds -127 to 127. +, -, *, %. min(), max(), sum(), avg(),
btree index and gin index are supported as other integer types.

Support
-------

This library is stored in an open [GitHub
repository](http://github.com/umitanuki/utinyint-postgresql).
Feel free to fork and contribute! Please file bug reports
via [GitHub Issues](http://github.com/umitanuki/utinyint-postgresql/issues/).

Author
------

[lxb]()

Copyright and License
---------------------

Copyright (c) lxb

This module is free software; you can redistribute it and/or modify it under
the [PostgreSQL License](http://www.opensource.org/licenses/postgresql).

Permission to use, copy, modify, and distribute this software and its 
documentation for any purpose, without fee, and without a written agreement is
hereby granted, provided that the above copyright notice and this paragraph
and the following two paragraphs appear in all copies.

In no event shall lxb be liable to any party for direct,
indirect, special, incidental, or consequential damages, including
lost profits, arising out of the use of this software and its documentation,
even if lxb has been advised of the possibility of such damage.

lxb specifically disclaim any warranties,
including, but not limited to, the implied warranties of merchantability and 
fitness for a particular purpose. The software provided hereunder is on an "as 
is" basis, and lxb has no obligations to provide maintenance,
support, updates, enhancements, or modifications.

