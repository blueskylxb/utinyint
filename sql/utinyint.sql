SET client_min_messages = warning;
\set ECHO none
\i utinyint.sql
\set ECHO all
RESET client_min_messages;

SELECT 0::utinyint;

SELECT 128::utinyint;
