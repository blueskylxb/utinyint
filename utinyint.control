# utinyint extension
comment = '1 byte integer data type'
default_version = '0.1.1'
module_pathname = '$libdir/utinyint'
relocatable = true
