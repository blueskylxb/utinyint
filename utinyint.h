
#ifndef utinyint_H
#define utinyint_H

#include <ctype.h>

#include "fmgr.h"


typedef unsigned char utinyint;

// the functions of SET_1_BYTE() and GET_1_BYTE() implement in include/postgres.h
#define DatumGetutinyint(X) ((utinyint) GET_1_BYTE(X)) 
#define utinyintGetDatum(X) ((Datum) SET_1_BYTE(X))

#define PG_GETARG_utinyint(n)	DatumGetutinyint(PG_GETARG_DATUM(n))
#define PG_RETURN_utinyint(x)	return utinyintGetDatum(x)

extern Datum utinyint_in(PG_FUNCTION_ARGS);
extern Datum utinyint_out(PG_FUNCTION_ARGS);
extern Datum utinyint_recv(PG_FUNCTION_ARGS);
extern Datum utinyint_send(PG_FUNCTION_ARGS);

extern Datum utinyint_to_i2(PG_FUNCTION_ARGS);
extern Datum i2_to_utinyint(PG_FUNCTION_ARGS);
extern Datum utinyint_to_i4(PG_FUNCTION_ARGS);
extern Datum i4_to_utinyint(PG_FUNCTION_ARGS);
extern Datum utinyint_to_bool(PG_FUNCTION_ARGS);
extern Datum bool_to_utinyint(PG_FUNCTION_ARGS);


#endif   /* utinyint_H */
