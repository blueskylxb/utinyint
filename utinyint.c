#include "postgres.h"

#include <ctype.h>
#include <limits.h>

#include "access/hash.h"
#include "catalog/pg_type.h"
#include "libpq/pqformat.h"
#include "utils/array.h"
#include "utinyint.h"
#include "utils/builtins.h"

#define SAMESIGN(a,b)	(((a) < 0) == ((b) < 0))

PG_MODULE_MAGIC;

PG_FUNCTION_INFO_V1(utinyint_in);
PG_FUNCTION_INFO_V1(utinyint_out);
PG_FUNCTION_INFO_V1(utinyint_recv);
PG_FUNCTION_INFO_V1(utinyint_send);

PG_FUNCTION_INFO_V1(utinyint_to_i2);
PG_FUNCTION_INFO_V1(i2_to_utinyint);
PG_FUNCTION_INFO_V1(utinyint_to_i4);
PG_FUNCTION_INFO_V1(i4_to_utinyint);
PG_FUNCTION_INFO_V1(utinyint_to_bool);
PG_FUNCTION_INFO_V1(bool_to_utinyint);


/*****************************************************************************
 *	 USER I/O ROUTINES														 *
 *****************************************************************************/

/*
 *	utinyint_in		- converts "num" to utinyint
 */
Datum
utinyint_in(PG_FUNCTION_ARGS)
{
	char	   *num = PG_GETARG_CSTRING(0);

	PG_RETURN_utinyint(pg_atoi(num, sizeof(utinyint), '\0'));
}

/*
 *	utinyint_out		- converts utinyint to "num"
 */
Datum
utinyint_out(PG_FUNCTION_ARGS)
{
	utinyint		arg1 = PG_GETARG_utinyint(0);
	char	   *result = (char *) palloc(4);	/* 3 digits, '\0' */

	pg_itoa(arg1, result);
	PG_RETURN_CSTRING(result);
}

/*
 *	utinyint_recv		- converts external binary format to utinyint
 */
Datum
utinyint_recv(PG_FUNCTION_ARGS)
{
	StringInfo	buf = (StringInfo) PG_GETARG_POINTER(0);

	PG_RETURN_utinyint((utinyint) pq_getmsgint(buf, sizeof(utinyint)));
}

/*
 * utinyint_send			- converts utinyint to binary format
 */
Datum
utinyint_send(PG_FUNCTION_ARGS)
{
	utinyint		arg1 = PG_GETARG_utinyint(0);
	StringInfoData buf;

	pq_begintypsend(&buf);
	pq_sendint(&buf, arg1, sizeof(utinyint));
	PG_RETURN_BYTEA_P(pq_endtypsend(&buf));
}

/*
 *		===================
 *		CONVERSION ROUTINES
 *		===================
 */

Datum
utinyint_to_i2(PG_FUNCTION_ARGS)
{
	utinyint		arg1 = PG_GETARG_utinyint(0);

	PG_RETURN_UINT16((uint16) arg1);
}

Datum
i2_to_utinyint(PG_FUNCTION_ARGS)
{
	uint16		arg1 = PG_GETARG_UINT16(0);

	if (arg1 < 0 || arg1 > UCHAR_MAX)
		ereport(ERROR,
				(errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),
				 errmsg("utinyint out of range")));

	PG_RETURN_utinyint((utinyint) arg1);
}

Datum
utinyint_to_i4(PG_FUNCTION_ARGS)
{
	utinyint		arg1 = PG_GETARG_utinyint(0);

	PG_RETURN_UINT32((uint32) arg1);
}

Datum
i4_to_utinyint(PG_FUNCTION_ARGS)
{
	uint32		arg1 = PG_GETARG_UINT32(0);

	if (arg1 < 0 || arg1 > UCHAR_MAX)
		ereport(ERROR,
				(errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),
				 errmsg("utinyint out of range")));

	PG_RETURN_utinyint((utinyint) arg1);
}

Datum
utinyint_to_bool(PG_FUNCTION_ARGS)
{
	if (PG_GETARG_utinyint(0) == 0)
		PG_RETURN_BOOL(false);
	else
		PG_RETURN_BOOL(true);
}

Datum
bool_to_utinyint(PG_FUNCTION_ARGS)
{
	if (PG_GETARG_BOOL(0) == false)
		PG_RETURN_utinyint(0);
	else
		PG_RETURN_utinyint(1);
}


