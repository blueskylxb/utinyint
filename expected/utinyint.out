SET client_min_messages = warning;
\set ECHO none
RESET client_min_messages;

SELECT 0::tinyint;
 tinyint 
---------
 0
(1 row)


SELECT 128::tinyint;
 tinyint 
---------
 128
(1 row)